contracts/adapters/SwapAdapter.sol

```diff
@@ -0,0 +1,230 @@
+        if (leftover > 0) {
+            payable(msg.sender).call{value: leftover}("");
+            // Do not revert if sender does not want to receive.
+        }
+    }
+
+    /**
+        @notice Function for depositing tokens, performing swap to ETH and bridging the ETH.
+        @param destinationDomainID  ID of chain deposit will be bridged to.
+        @param recipient Recipient of the deposit.
+        @param  token Output token to be deposited after swapping.
+        @param amountOutMinimum Minimal amount of tokens to be accepted as a swap output.
+        @param pathTokens Addresses of the tokens for Uniswap swap. WETH address is used for ETH.
+        @param pathFees Fees for Uniswap pools.
+    */
+   function depositEthToTokens(
+        uint8 destinationDomainID,
+        address recipient,
+        address token,
+        uint256 amountOutMinimum,
+        address[] memory pathTokens,
+        uint24[] memory pathFees
+    ) external payable {
+        bytes32 resourceID = tokenToResourceID[token];
+        if (resourceID == bytes32(0)) revert TokenInvalid();
+
+        // Compose depositData
+        bytes memory depositDataAfterAmount = abi.encodePacked(
+            uint256(20),
+            recipient
+        );
+        if (msg.value == 0) revert InsufficientAmount(msg.value);
+        uint256 fee;
+        {
+            address feeHandlerRouter = _bridge._feeHandler();
+            (fee, ) = IFeeHandler(feeHandlerRouter).calculateFee(
+                address(this),
+                _bridge._domainID(),
+                destinationDomainID,
+                resourceID,
+                abi.encodePacked(msg.value, depositDataAfterAmount),
+                ""  // feeData - not parsed
+            );
+        }
+
+        if (msg.value < fee) revert MsgValueLowerThanFee(msg.value);
+        uint256 amountOut;
+        {
+            uint256 swapAmount = msg.value - fee;
+            // Convert everything except the fee
+
+            // Swap ETH to tokens (exact input)
+            bytes memory path = _verifyAndEncodePath(
+                pathTokens,
+                pathFees,
+                _weth,
+                token 
+            );
+            IV3SwapRouter.ExactInputParams memory params = IV3SwapRouter.ExactInputParams({
+                path: path,
+                recipient: address(this),
+                amountIn: swapAmount,
+                amountOutMinimum: amountOutMinimum
+            });
+
+            amountOut = _swapRouter.exactInput{value: swapAmount}(params);
+            emit TokensSwapped(token, amountOut);
+        }
+
+        bytes memory depositData = abi.encodePacked(
+            amountOut,
+            depositDataAfterAmount
+        );
+
+        address ERC20HandlerAddress = _bridge._resourceIDToHandlerAddress(resourceID);
+        IERC20(token).safeApprove(address(ERC20HandlerAddress), amountOut);
+        _bridge.deposit{value: fee}(destinationDomainID, resourceID, depositData, "");
```

Comment:
same as above, we need to be able to pass data for optional gmp call with transfer. This also needs to be accounted in fee handler